src/solver/expressions/visitors/AstDOTStyleVisitor.cpp

```diff
@@ -0,0 +1,251 @@
+}
+
+void AstDOTStyleVisitor::visit(const Nodes::MultiplicationNode* node, std::ostream& os)
+{
+    processBinaryOperation(node, "*", NodeStyle::BinaryStyle, os);
+}
+
+void AstDOTStyleVisitor::visit(const Nodes::DivisionNode* node, std::ostream& os)
+{
+    processBinaryOperation(node, "/", NodeStyle::BinaryStyle, os);
+}
+
+void AstDOTStyleVisitor::visit(const Nodes::EqualNode* node, std::ostream& os)
+{
+    processBinaryOperation(node, "==", NodeStyle::ComparisonStyle, os);
+}
+
+void AstDOTStyleVisitor::visit(const Nodes::LessThanOrEqualNode* node, std::ostream& os)
+{
+    processBinaryOperation(node, "<=", NodeStyle::ComparisonStyle, os);
+}
+
+void AstDOTStyleVisitor::visit(const Nodes::GreaterThanOrEqualNode* node, std::ostream& os)
+{
+    processBinaryOperation(node, ">=", NodeStyle::ComparisonStyle, os);
+}
+
+void AstDOTStyleVisitor::visit(const Nodes::VariableNode* node, std::ostream& os)
+{
+    auto id = getNodeID(node);
+    emitNode(id, "Var(" + node->value() + ")", NodeStyle::VariableStyle, os);
+}
+
+void AstDOTStyleVisitor::visit(const Nodes::ParameterNode* node, std::ostream& os)
+{
+    auto id = getNodeID(node);
+    emitNode(id, "Param(" + node->value() + ")", NodeStyle::ParameterStyle, os);
+}
+
+void AstDOTStyleVisitor::visit(const Nodes::LiteralNode* node, std::ostream& os)
+{
+    auto id = getNodeID(node);
+    emitNode(id, std::to_string(node->value()), NodeStyle::LiteralStyle, os);
+}
+
+void AstDOTStyleVisitor::visit(const Nodes::NegationNode* node, std::ostream& os)
+{
+    auto id = getNodeID(node);
+    emitNode(id, "-", NodeStyle::NegationStyle, os);
+    auto childId = getNodeID(node->child());
+    os << "  " << id << " -> " << childId << ";\n";
+    dispatch(node->child(), os);
+}
+
+void AstDOTStyleVisitor::visit(const Nodes::PortFieldNode* node, std::ostream& os)
+{
+    auto id = getNodeID(node);
+    emitNode(id,
+             "PF(" + node->getPortName() + "," + node->getFieldName() + ")",
+             NodeStyle::PortFieldStyle,
+             os);
+}
+
+void AstDOTStyleVisitor::visit(const Nodes::ComponentVariableNode* node, std::ostream& os)
+{
+    auto id = getNodeID(node);
+    emitNode(id,
+             "CV(" + node->getComponentId() + "," + node->getComponentName() + ")",
+             NodeStyle::ComponentVariableStyle,
+             os);
+}
+
+void AstDOTStyleVisitor::visit(const Nodes::ComponentParameterNode* node, std::ostream& os)
+{
+    auto id = getNodeID(node);
+    emitNode(id,
+             "CP(" + node->getComponentId() + "," + node->getComponentName() + ")",
+             NodeStyle::ComponentParameterStyle,
+             os);
+}
+
+std::string AstDOTStyleVisitor::name() const
+{
+    return "AstGraphVisitor";
+}
+
+unsigned int AstDOTStyleVisitor::getNodeID(const Nodes::Node* node)
+{
+    const auto& node_name = node->name();
+    if (nodeIds_.find(node_name) == nodeIds_.end())
+    {
+        nodeIds_[node->name()][node] = ++nodeCount_;
```

Comment:
It seems like you don't need to mix node IDs and category counting, you might consider two separate maps here